### PR TITLE
Remove closed world validation checks

### PR DESCRIPTION
**Pull request by kripken**
These were added to avoid common problems with closed world mode, but
in practice they are causing more harm than good, forcing users to work
around them. In the meantime (until #6965 ), remove this validation to unblock
current toolchain makers.

There is a risk of getting more user reports here, but I don't see any other
risk. @tlively you were worried about fuzz bugs, but I don't think we depend
on validation for deciding what to optimize, so that should be ok? (I've fuzzed
a few thousand iterations so far without issue, too.)

### CODE REVIEW

**Review by tlively — APPROVED** [review]
🎉 I'm pleasantly surprised that this doesn't cause issues 👍

**Comment by kripken** [comment]
Added a similar fix for abstract type refining (found not by fuzzing but analytically).

File: test/lit/passes/abstract-type-refining.wast [hunk 1]
```diff
@@ -1304,3 +1304,55 @@
     )
   )
 )
+
+;; $A is never created, but $B is, so all appearances of $A, like in the cast
+;; and the struct field, can be replaced by $B. However, these are public types,
+;; which we must not modify, so we do not.
```
Comment (tlively) [inline test/lit/passes/abstract-type-refining.wast:1310]:
```suggestion
;; $A is never created, but $B is, so all appearances of $A, like in the cast
;; and the struct field, could be replaced by $B, except that $A is a public type,
;; so might be created outside the module.
```

(assuming we don't make changes to the relevant code)

File: src/passes/AbstractTypeRefining.cpp [hunk 2]
```diff
@@ -106,6 +106,12 @@ struct AbstractTypeRefining : public Pass {
       }
     }
 
+    // Assume all public types are created, which makes them non-abstract and
+    // hence ignored below.
+    for (auto type : ModuleUtils::getPublicHeapTypes(*module)) {
+      createdTypes.insert(type);
+    }
```
Comment (tlively) [inline src/passes/AbstractTypeRefining.cpp:113]: Do we need to make this assumption? If we assumed the public types are not allocated outside the module due to closed-world, we could still optimize the casts in the test case without changing the public types. It is probably safer to never make assumptions about public types, though 🤔

**Review by tlively — COMMENTED** [review]
Code

File: src/passes/AbstractTypeRefining.cpp (hunk 2 above)
Comment (kripken) [reply src/passes/AbstractTypeRefining.cpp:113]: Yeah, we could do that, if we checked and found we'd need no changes to public types. I'll add a TODO, but it seems nontrivial.

**Comment by kripken** [comment]
No new fuzzer issues, landing.

**Pull request merged**